tiny_dnn/layers/layer.h

```diff
@@ -347,6 +353,70 @@ class layer : public node {
     return *this;
   }
 
+  /**
+   * @name Parameter Getters and Setters
+   * @{
+   */
+
+  /** @brief Add new parameter to this layer, to be called in constructor. */
+  void add_parameter(size_t width,
+                     size_t height,
+                     size_t depth,
+                     size_t n_fmaps,
+                     parameter_type type,
+                     bool trainable = true) {
+    parameters_.push_back(
+      new parameter(width, height, depth, n_fmaps, type, trainable));
+  }
+
+  /**
+   * @brief Get pointers to parameters of this layer.
+   *
+   * This can be used during training phase when parameter update is
+   * required.
+   *
+   * @param trainable_only flag to return only the trainable parameters.
+   * @return std::vector of pointers to parameters
+   */
+  std::vector<parameter *> parameters(bool trainable_only = false) {
```

Comment:
why not using `shared_ptr` ? could you run a small benchmark to see if there's any argument in favour/against ?